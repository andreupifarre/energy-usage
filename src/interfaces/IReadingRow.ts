export interface IReadingRow {
  readingDate: string;
  cumulative: number;
  unit: string;
}
