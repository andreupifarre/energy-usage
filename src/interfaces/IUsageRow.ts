export interface IUsageRow {
  date: string;
  energyUsage: number;
}
