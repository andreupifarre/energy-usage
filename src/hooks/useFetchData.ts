import { useState, useEffect } from 'react';
import { IReadingRow } from '../interfaces/IReadingRow';

const useFetchData = () => {
  const [results, setResults] = useState<IReadingRow[]>([]);

  useEffect(() => {
    const url = `https://storage.googleapis.com/bulb-interview/meterReadingsReal.json`;
    fetch(url)
      .then(response => response.json())
      .then(data => setResults(data.electricity));
  }, []);

  return results;
};

export default useFetchData;
