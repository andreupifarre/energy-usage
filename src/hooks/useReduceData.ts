import * as moment from 'moment';
import { IReadingRow } from '../interfaces/IReadingRow';
import { IUsageRow } from '../interfaces/IUsageRow';

const useReduceData = (readings: IReadingRow[]) => {
  const energyUsageData: IUsageRow[] = [];

  const lerp = (x0, x1, y0, y1) => {
    var slope = (y1 - y0) / (x1 - x0);
    return x => (x - x0) * slope + y0;
  };

  let prevMonth = 0;
  for (let i = 1; i < readings.length - 1; i++) {
    const x0 = moment(readings[i - 1].readingDate).date() + 31 * (i - 1);
    const x1 = moment(readings[i].readingDate).date() + 31 * i;
    const y0 = readings[i - 1].cumulative;
    const y1 = readings[i].cumulative;

    const endMonthReadings = lerp(x0, x1, y0, y1)(31 * i);
    const energyUsage = parseFloat((endMonthReadings - prevMonth).toFixed(2));

    if (i === 1) {
      prevMonth = endMonthReadings;
      continue;
    }

    const month = energyUsageData.push({
      date: moment(readings[i].readingDate).format('MMMM-YYYY') + '',
      energyUsage: energyUsage
    });

    prevMonth = endMonthReadings;
  }

  return energyUsageData;
};

export default useReduceData;
