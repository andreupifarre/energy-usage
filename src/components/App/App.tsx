import * as React from 'react';
import useFetchData from '../../hooks/useFetchData';
import useReduceData from '../../hooks/useReduceData';
import { IReadingRow } from '../../interfaces/IReadingRow';
import { IUsageRow } from '../../interfaces/IUsageRow';
import ReadingsTable from '../ReadingsTable/ReadingsTable';
import UsageChart from '../UsageChart/UsageChart';
import * as styles from './App.scss';

const App = () => {
  const readings: IReadingRow[] = useFetchData();
  const usage: IUsageRow[] = useReduceData(readings);

  return (
    <div className={styles.App}>
      <h2>Energy Usage</h2>
      <UsageChart usage={usage} />
      <h2>Meter Readings</h2>
      <ReadingsTable readings={readings} />
    </div>
  );
};

export default App;
