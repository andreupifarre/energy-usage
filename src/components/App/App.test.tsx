import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import App from './App';
import { IReadingRow } from '../../interfaces/IReadingRow';
import { IUsageRow } from '../../interfaces/IUsageRow';
import * as useReduceData from '../../hooks/useReduceData';
import * as useFetchData from '../../hooks/useFetchData';

jest.mock('../../hooks/useReduceData');
jest.mock('../../hooks/useFetchData');

const mockUseReduceData = jest.spyOn(useReduceData, 'default');
const mockUseFetchData = jest.spyOn(useFetchData, 'default');

const wrapper = () => shallow(<App />);

describe('App', () => {
  it('should render without crashing', () => {
    const fetchedData: IReadingRow[] = [
      {
        cumulative: 17580,
        readingDate: '2017-03-28T00:00:00.000Z',
        unit: 'kWh'
      }
    ];

    const reducedData: IUsageRow[] = [
      {
        date: 'March-2017',
        energyUsage: 17580
      }
    ];

    mockUseFetchData.mockReturnValue(fetchedData);
    mockUseReduceData.mockReturnValue(reducedData);

    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should render correctly', () => {
    expect(wrapper().getElements()).toMatchSnapshot();
  });
});
