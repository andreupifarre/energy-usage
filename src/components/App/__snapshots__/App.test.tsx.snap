// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
Array [
  <div
    className="App"
  >
    <h2>
      Energy Usage
    </h2>
    <UsageChart
      usage={
        Array [
          Object {
            "date": "March-2017",
            "energyUsage": 17580,
          },
        ]
      }
    />
    <h2>
      Meter Readings
    </h2>
    <ReadingsTable
      readings={
        Array [
          Object {
            "cumulative": 17580,
            "readingDate": "2017-03-28T00:00:00.000Z",
            "unit": "kWh",
          },
        ]
      }
    />
  </div>,
]
`;
